blog: update the content for clarity (#93)
pingcap/blog: 2016-11-17-mvcc-in-tikv.md

## Introduction to concurrency control

> 

Author: Replace this sentence with the concept of transaction to be more logic and relevant in the context.

> In a database system, a transaction might consist of one or more data-manipulation statements and queries, each reading and/or writing information in the database. Users of database systems consider consistency and integrity of data as highly important. However, concurrency controlling problems like data races must be resolved since there will be a lot of connections to the database. Owing to this background, the DBMS must ensure that the resulting concurrent access patterns are safe and ideally serializable.

Reviewer: I think this paragraph mismatches the technical depth of the rest of the article. Can we maybe shorten it to something close to the last sentence, and say "in this post I will explore the complexities of how to implement such functionality"?

Author: Agreed.

Author: @morgo I removed this paragraph entirely and add the notion of serializability to the beginning of next paragraph.

Reviewer: LGTM

> *Multi-Version Concurrency Control (MVCC)* is an elegant solution for this problem, where each update creates a new version of the data object instead of updating it in-place, so that concurrenct readers can still see the old version while the update transaction proceeds. Such a strategy can prevent read-only transactions from waiting. In fact, locking is not required at all. This is an extremely desirable property and the reason why many database systems like PostgreSQL, Oracle, and Microsoft SQL Server implement MVCC.

Reviewer 2: ```suggestion
*Multi-Version Concurrency Control (MVCC)* is an elegant solution for this problem, where each update creates a new version of the data object instead of updating it in-place, so that concurrent readers can still see the old version while the update transaction proceeds. Such a strategy can prevent read-only transactions from waiting. In fact, locking is not required at all. This is an extremely desirable property and the reason why many database systems like PostgreSQL, Oracle, and Microsoft SQL Server implement MVCC.
```

In this post, we will explore the complexities of implementation of MVCC in TiKV.

## MVCC in TiKV

Let's dive into `TiKV`'s MVCC implementation, which is located at [src/storage](https://github.com/tikv/tikv/blob/1050931de5d9b47423f997d6fc456bd05bd234a7/src/storage/mod.rs).

### Timestamp Oracle(TSO)

> Since `TiKV` is a distributed storage system, it needs a globally unique time service, called `Timestamp Oracle`(TSO), to allocate a monotonic increasing timestamp. Similar to TrueTime API` from Google's [Spanner](http://static.googleusercontent.com/media/research.google.com/en//archive/spanner-osdi2012.pdf), this service is implemented in Placement Driver (PD) in TiKV. Every `TS` represents a monotonic increasing timestamp.

Author: ```suggestion
Since `TiKV` is a distributed storage system, it needs a globally unique time service, called `Timestamp Oracle`(TSO), to allocate a monotonic increasing timestamp. Similar to the TrueTime API from Google's [Spanner](http://static.googleusercontent.com/media/research.google.com/en//archive/spanner-osdi2012.pdf), this service is implemented in Placement Driver (PD) in TiKV. Every `TS` represents a monotonic increasing timestamp.
```

> Since `TiKV` is a distributed storage system, it needs a globally unique time service, called `Timestamp Oracle`(TSO), to allocate a monotonic increasing timestamp. This function is provided in Placement Driver (PD) in TiKV, which is inspired by `TrueTime API` from Google's [Spanner](http://static.googleusercontent.com/media/research.google.com/en//archive/spanner-osdi2012.pdf). Every `TS` represents a monotonic increasing timestamp.

Author: ```suggestion
Since `TiKV` is a distributed storage system, it needs a globally unique time service, called `Timestamp Oracle`(TSO), to allocate a monotonic increasing timestamp. Similar to TrueTime API` from Google's [Spanner](http://static.googleusercontent.com/media/research.google.com/en//archive/spanner-osdi2012.pdf), this function is implemented in Placement Driver (PD) in TiKV. Every `TS` represents a monotonic increasing timestamp.
```

}
```

This `start` function in the example above explains how the storage struct runs.

### Engine

> [Engine](https://github.com/pingcap/tikv/blob/master/src/storage/engine/mod.rs#L44) is the trait which describes the actual database used in storage system. It is implemented in [raftkv](https://github.com/pingcap/tikv/blob/master/src/storage/engine/raftkv.rs#L91) and [Enginerocksdb](https://github.com/tikv/tikv/blob/master/src/storage/engine/rocksdb_engine.rs#L240).

Reviewer 2: ```suggestion
[Engine](https://github.com/pingcap/tikv/blob/master/src/storage/engine/mod.rs#L44) is the trait which describes the actual database used in storage system. It is implemented in [raftkv](https://github.com/pingcap/tikv/blob/master/src/storage/engine/raftkv.rs#L91) and [rocksdb_engine](https://github.com/tikv/tikv/blob/master/src/storage/engine/rocksdb_engine.rs#L240).
```

> [Engine](https://github.com/pingcap/tikv/blob/master/src/storage/engine/mod.rs#L44) is the trait which describes the actual database used in storage system. It is implemented in [raftkv](https://github.com/pingcap/tikv/blob/master/src/storage/engine/raftkv.rs#L91) and [rocksdb_engine](https://github.com/tikv/tikv/blob/master/src/storage/engine/rocksdb_engine.rs#L240).

Author: ```suggestion
[Engine](https://github.com/pingcap/tikv/blob/master/src/storage/engine/mod.rs#L44) is the trait which describes the actual database used in the storage system. It is implemented in [raftkv](https://github.com/pingcap/tikv/blob/1050931de5d9b47423f997d6fc456bd05bd234a7/src/storage/engine/raftkv.rs#L91) and [rocksdb_engine](https://github.com/tikv/tikv/blob/1050931de5d9b47423f997d6fc456bd05bd234a7/src/storage/engine/rocksdb_engine.rs#L240).
```

### StorageHandle

> `StorageHandle` is the struct that handles commands received from `sendch`. The IO is processed by [`mio`](https://github.com/carllerche/mio).

Reviewer 3: ```suggestion
`StorageHandle` is the struct that handles commands received from `sendch`. The I/O is processed by [`mio`](https://github.com/carllerche/mio).
```

Then functions like `async_get` and `async_batch_get` in the `storage` struct will send the corresponding commands to the channel, which can be obtained by the scheduler to execute asynchronously.

> The MVCC layer is called in [Scheduler](https://github.com/pingcap/tikv/blob/master/src/storage/txn/scheduler.rs#L763).

Author: ```suggestion
The MVCC layer is called in [Scheduler](https://github.com/pingcap/tikv/blob/1050931de5d9b47423f997d6fc456bd05bd234a7/src/storage/txn/scheduler.rs#L763).
```

> The storage receives commands from clients and sends commands as messages to the scheduler. Then the scheduler will process the command or [call corresponding asynchronous function](https://github.com/pingcap/tikv/blob/master/src/storage/txn/scheduler.rs#L643). There are two types of operations - read and write. Read is implemented in [MvccReader](https://github.com/tikv/tikv/blob/1050931de5d9b47423f997d6fc456bd05bd234a7/src/storage/mvcc/reader.rs#L20), which is easy to understand so we will not elaborate on it. Let's focus on write, which is the core of MVCC implementation.

Author: ```suggestion
The storage receives commands from clients and sends commands as messages to the scheduler. Then the scheduler will process the command or [call corresponding asynchronous function](https://github.com/pingcap/tikv/blob/1050931de5d9b47423f997d6fc456bd05bd234a7/src/storage/txn/scheduler.rs#L643). There are two types of operations - read and write. Read is implemented in [MvccReader](https://github.com/tikv/tikv/blob/1050931de5d9b47423f997d6fc456bd05bd234a7/src/storage/mvcc/reader.rs#L20), which is easy to understand so we will not elaborate on it. Let's focus on write, which is the core of MVCC implementation.
```

### MVCC

> #### Column family

Author: **Column family**  content moved from below.

> Compared with Percolator where the information such as Lock is stored by adding extra column to a specific row, TiKV uses a column family (CF) in RocksDB to handle all the information related to Lock. To be specific, TiKV stores the Key-Values, Locks and Writes information in `CF_DEFAULT`, `CF_LOCK`, `CF_WRITE`.

Reviewer 3: ```suggestion
Compared with Percolator where the information such as Lock is stored by adding an extra column to a specific row, TiKV uses a column family (CF) in RocksDB to handle all the information related to Lock. To be specific, TiKV stores the Key-Values, Locks and Writes information in `CF_DEFAULT`, `CF_LOCK`, and `CF_WRITE`.
```

All the values of the CF are encoded as following:

> |  | default | lock | write |

Reviewer 3: ```suggestion
| Default | Lock | Write |
```

> | Default | Lock | Write |

Reviewer 2: There is something wrong with the table layout.

> | Default | Lock | Write |

Author: ```suggestion
| | Default | Lock | Write |
```

| --- | --- | --- | --- |

> | **key** | z{encoded_key}{start_ts(desc)} | z{encoded_key} | z{encoded_key}{commit_ts(desc)} |

Reviewer 3: ```suggestion
| **Key** | z{encoded_key}{start_ts(desc)} | z{encoded_key} | z{encoded_key}{commit_ts(desc)} |
```

> | **value** | {value} | {flag}{primary_key}{start_ts(varint)} | {flag}{start_ts(varint)} |

Reviewer 3: ```suggestion
| **Value** | {value} | {flag}{primary_key}{start_ts(varint)} | {flag}{start_ts(varint)} |
```

More details can be found [here](https://github.com/pingcap/tikv/issues/1077).

> 

Author: Removed the Rollback topic since it has been introduced in the PreWrite phase.

#### Transaction model

Here comes the core of the transaction model for TiKV, which is MVCC powered by 2-phase commit. There are two stages in one transaction:

- **Prewrite**

> 

Author: Removed the topic of **TiKV-Ctl for MVCC** since this feature has evolved a lot since this article.

> 

Author: Removed the topic of **TiKV-Ctl for MVCC** since this feature has evolved a lot since this article.

  1. The transaction starts. The client obtains the current timestamp (`startTS`) from TSO.
  2. Select one row as the primary row, the others as the secondary rows.

>   3. Check [whether there is other locks on this row](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L71) or whether there are some commits located after `startTS`. These two situations will lead to conflicts. If either happens, the commit fails and [rollback](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L115) will be called.

Reviewer 3: ```suggestion
  3. Check [whether there is another lock on this row](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L71) or whether there are any commits located after `startTS`. These two situations will lead to conflicts. If either happens, the commit fails and [rollback](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L115) will be called.
```

>   3. Check [whether there is another lock on this row](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L71) or whether there are any commits located after `startTS`. These two situations will lead to conflicts. If either happens, the commit fails and [rollback](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L115) will be called.

Author: ```suggestion
  3. Check [whether there is another lock on this row](https://github.com/pingcap/tikv/blob/1050931de5d9b47423f997d6fc456bd05bd234a7/src/storage/mvcc/txn.rs#L71) or whether there are any commits located after `startTS`. These two situations will lead to conflicts. If either happens, the commit fails and [rollback](https://github.com/pingcap/tikv/blob/1050931de5d9b47423f997d6fc456bd05bd234a7/src/storage/mvcc/txn.rs#L115) will be called.
```

>   4. [Lock the primary row](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L80).

Author: ```suggestion
  4. [Lock the primary row](https://github.com/pingcap/tikv/blob/1050931de5d9b47423f997d6fc456bd05bd234a7/src/storage/mvcc/txn.rs#L80).
```

  5. Repeat the steps above on secondary rows.

> 

Author: The CF related info is moved before the 2PC process since the CF_WRITE column is referenced in the Commit phase.

- **Commit**

  1. Obtain the commit timestamp `commit_ts` from TSO.
  2. Check whether the lock on the primary row still exists. Proceed if the lock is still there. Roll back if not.
  3. Write to column `CF_WRITE` with `commit_ts`.
  4. Clear the corresponding primary lock.
  5. Repeat the steps above on secondary rows.

#### Garbage collector

> It is easy to predict that there will be more and more MVCC versions if there is no [Garbage Collector](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L143) to remove the invalid versions. But we cannot simply remove all the versions before a safe point, for there may be only one version for a key, which must be kept. In TiKV, if there is any `Put` or `Delete` records before the safe point, then all the latter writes can be deleted, otherwise only `Delete`, `Rollback` and `Lock` will be deleted.

Reviewer 3: ```suggestion
It is easy to predict that there will be more and more MVCC versions if there is no [Garbage Collector](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L143) to remove the invalid versions. But we cannot simply remove all the versions before a safe point, for there may be only one version for a key, which must be kept. In TiKV, if there is any `Put` or `Delete` records before the safe point, then all the latter writes can be deleted; otherwise only `Delete`, `Rollback` and `Lock` will be deleted.
```

> It is easy to predict that there will be more and more MVCC versions if there is no [Garbage Collector](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L143) to remove the invalid versions. But we cannot simply remove all the versions before a safe point, for there may be only one version for a key, which must be kept. In TiKV, if there is any `Put` or `Delete` records before the safe point, then all the latter writes can be deleted; otherwise only `Delete`, `Rollback` and `Lock` will be deleted.

Author: ```suggestion
It is easy to predict that there will be more and more MVCC versions if there is no [Garbage Collector](https://github.com/pingcap/tikv/blob/1050931de5d9b47423f997d6fc456bd05bd234a7/src/storage/mvcc/txn.rs#L143) to remove the invalid versions. But we cannot simply remove all the versions before a safe point, for there may be only one version for a key, which must be kept. In TiKV, if there is any `Put` or `Delete` records before the safe point, then all the latter writes can be deleted; otherwise only `Delete`, `Rollback` and `Lock` will be deleted.
```
